CARDS-2171: Refactor FilterServlet to provide all the information needed by Filters.jsx

veronikaslc opened the pull request:
This refactoring affects the dashboard filters - test they still work as expected.

Some visible changes were introduced:
* Filters variable autocomplete:
  * Each option has a "sections breadcrumbs" (list of the labels of all ancestor sections) as a secondary label
  * Once the option is selected, the helper text of the input becomes
  `"<Parent questionnaire title>[: <section breadcrumbs>]"`
* http://localhost:8080/content.html/Forms?questionnaire=/Questionnaires/<QUESTIONNAIRE_ID> now works

marta- committed:
CARDS-2171: Refactor FilterServlet to provide all the information needed by Filters.jsx
Added the questionnaireTitle field to each question returned by FilterServlet

veronikaslc committed:
CARDS-2171: Refactor FilterServlet to provide all the information needed by Filters.jsx

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/FilterServlet.java:
```diff
@@ -45,7 +45,12 @@
  * </p>
  * <ul>
- * <li><code>questionnaire</code>: a path to a questionnaire whose filterable options to retrieve</li>
+ * <li><code>questionnaire</code>: a path to a questionnaire whose filterable options to retrievec
+ * <li>if absent, by default generating all filters from all questionnaires<li>
```
```suggestion
 * <li><code>questionnaire</code>: a path to a questionnaire whose filterable options to retrieve; if absent, retrieve all filterable options from all questionnaires</li>
```

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/FilterServlet.java:
```diff
@@ -45,7 +45,12 @@
- * <li><code>questionnaire</code>: a path to a questionnaire whose filterable options to retrieve</li>
+ * <li><code>questionnaire</code>: a path to a questionnaire whose filterable options to retrievec
+ * <li>if absent, by default generating all filters from all questionnaires<li>
+ * <li><code>include</code>: a parameter include that allows to specify which kinds of filters to retrieve</li>
+ * <li>by default both metadata filters and question filters are retrieved</li>
+ * <li><code>include=metadata</code> returns only metadata filters</li>
+ * <li><code>include=questions</code> returns only question filters</li>
```
```suggestion
 * <li><code>include</code>: a parameter include that allows to specify which kinds of filters to retrieve. By default, both metadata filters and question filters are retrieved; <code>include=metadata</code> returns only metadata filters, <code>include=questions</code> returns only question filters</li>
```

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/FilterServlet.java:
```diff
@@ -60,23 +65,22 @@ public class FilterServlet extends SlingSafeMethodsServlet
-            ? getAllFieldsFromAllQuestionnaires(request.getResourceResolver(), homepagePath)
-            : getAllFields(request.getResourceResolver(), questionnaire);
+            ? getAllFieldsFromAllQuestionnaires(request.getResourceResolver(), homepagePath, include)
+            : getAllFields(request.getResourceResolver(), homepagePath + "/" + questionnaire, include);
```
The javadoc says `questionnaire` is a path to a questionnaire resource, and the old code operates under the assumption that this parameter holds the full path. Why is it appended to the homepage path in the new code?

veronikaslc replied:
Frankly speaking I don't even seem to find the `old code` that is really using it from a request point of side.
In other words, where do we exactly **define** the `questionnaire` param to be later set as a request part now?

You can trace it through the `LiveTable` up to `Forms.jsx`, but [that line](https://github.com/data-team-uhn/cards/commit/86b67f97b3432befed16c4bbe74c7c1ca61b2f84#diff-1c963290908c6c6a84bea0f42e36cabc93a537b6ef117d4df547f3ede61d910dR25) is a residual of design dated 4 years ago when LFS-170 was implemented. It does not seem that we have the individual questionnaires view anymore.

Separately thinking about parameterization, using questionnaire `"@path"` as a param for `questionnaire` feels more logical that full path.

marta- replied:
The `@path` of a questionnaire _is_ its full path and includes `/Questionnaires` which is the `homepagePath` here.

I removed the `homepagePath/` prefix in c74770f.

Historically, the `questionnaire` parameter was used when we used to have a dashboard box for each questionnaire, which we could expand to a fullscreen `FormView` listing of all the forms for that particular questionnaire. It's no longer used in the UI anywhere at this time.

marta- replied:
Updated `FormView` process the `questionnaire` url parameter as a path instead of a uuid and pass it on to the `Filters` accordingly in edccf08.

Now this works:
![image](https://user-images.githubusercontent.com/651980/233495549-0af83647-dce9-4997-89d4-7495b68f56b4.png)

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/FilterServlet.java:
```diff
@@ -60,23 +65,22 @@ public class FilterServlet extends SlingSafeMethodsServlet
 
-        // Return the entire thing as a json file, except join together fields that have the same
-        // name and type
+        // Return the entire thing as a json file, except join together fields that have the same name and type
```
We're no longer joining together fields that have the same name and type

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/FilterServlet.java:
```diff
@@ -86,20 +90,29 @@ public void doGet(final SlingHttpServletRequest request, final SlingHttpServletR
-        this.copyQuestions(resource.adaptTo(JsonObject.class), builder);
+
+        // Generate the metadata filters
+        if (StringUtils.isBlank(include) || "metadata".equals(include)) {
+            builder.add("metadataFilters", this.addMetadatalFilters(false));
+        }
+
+        if (StringUtils.isBlank(include) || "questions".equals(include)) {
```
I think the interpretation of the "include" parameter should be done directly in `doGet`.

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/FilterServlet.java:
```diff
@@ -60,23 +65,22 @@ public class FilterServlet extends SlingSafeMethodsServlet
     {
         // Is there a questionnaire specified?
         String questionnaire = request.getParameter("questionnaire");
+        String include = request.getParameter("include");
```
The interpretation of the `include` parameter fits better up here than in both `getAllFieldsFromAllQuestionnaires` and `getAllFields`:

```
boolean includeMetadata = StringUtils.isBlank(include) || "metadata".equals(include);
boolean includeQuestions = StringUtils.isBlank(include) || "questions".equals(include);

JsonObjectBuilder builder = Json.createObjectBuilder();

if (includeMetadata) {
    builder.add("metadataFilters", addMetadataFilters(questionnaire == null));
}

if (includeQuestions) {
   ...
}
```

veronikaslc replied:
We can not move up processing "questions". There is no `results` to process in case of `if (includeQuestions) {` up in `doGet` for the `questionnaire == null` case. 
Moving up everything basically cancels the need of 2 methods `getAllFields` and `getAllFieldsFromAllQuestionnaires`, which overloads `doGet`. I find those methods useful as they make process more clear.

marta- replied:
We can, we just need to call those 2 functions inside an `if (includeQuestions) {}`  instead of passing `includeQuestions` to them as parameters. I'll make a commit as it's easier than explaining all the details in words.

marta- replied:
Addressed in c74770f

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/FilterServlet.java:
```diff
@@ -109,129 +122,125 @@ private JsonObject getAllFields(ResourceResolver resolver, String questionnaireP
      */
-    private void copyFields(JsonObject question, String key, JsonObjectBuilder builder, Map<String, String> seenTypes,
-            Map<String, String> seenElements)
+    private void copyQuestions(JsonObject datum, JsonArrayBuilder questions, JsonArrayBuilder sectionBreadcrumbs)
```
**Bug**: All the labels of all the sections end up being associated with every questions, not just the parent sections of that question. See for example any question from the CPESIC questionnaire:

![image](https://user-images.githubusercontent.com/651980/232903180-4415cd52-dbd9-47df-b4d4-d517224b579f.png)

marta- replied:
Fixed in c74770f

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/FilterServlet.java:
```diff
@@ -109,129 +122,125 @@ private JsonObject getAllFields(ResourceResolver resolver, String questionnaireP
-     */
-    private void copyQuestions(JsonObject questions, JsonObjectBuilder builder, Map<String, String> seenTypes,
-            Map<String, String> seenElements)
+    private JsonArrayBuilder addMetadatalFilters(final boolean addQuestionnaireFilter)
```
Typo in the name
```suggestion
    private JsonArrayBuilder addMetadataFilters(final boolean addQuestionnaireFilter)
```

veronikaslc committed:
CARDS-2171: Refactor FilterServlet to provide all the information needed by Filters.jsx

Better documentation

Co-authored-by: Marta Girdea <marta@phenotips.org>

veronikaslc committed:
CARDS-2171: Refactor FilterServlet to provide all the information needed by Filters.jsx

Better documentation

Co-authored-by: Marta Girdea <marta@phenotips.org>

veronikaslc committed:
CARDS-2171: Refactor FilterServlet to provide all the information needed by Filters.jsx
Fixed typo

sashaandjic commented:
![image](https://user-images.githubusercontent.com/23243244/233216541-f908ac38-38d0-4a61-b856-2044554b0bc6.png)

veronikaslc committed:
CARDS-2171: Refactor FilterServlet to provide all the information needed by Filters.jsx
Fixed javadoc string length

veronikaslc commented:
@sashaandjic Please **do not test** if there is no "Test me" label on PR.

veronikaslc committed:
CARDS-2171: Refactor FilterServlet to provide all the information needed by Filters.jsx
Moved the interpretation of the include parameter to be done directly in doGet method

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/FilterServlet.java:
```diff
@@ -60,23 +65,33 @@ public class FilterServlet extends SlingSafeMethodsServlet
+
+        // Generate the metadata filters
+        if (includeMetadata) {
+            builder.add("metadataFilters", this.addMetadataFilters(false));
```
**Bug**: now the "Questionnaire" filter is never added because the parameter of `addMetadataFilters` is always false.

**Code style**:
* method naming: `add` is suited for actively modifying either `this` or one of the method's parameters; when we just return something without altering any state we should use `get`, `generate`, etc.
* we don't need `this.` in front of methods.

marta- replied:
Addressed both concerns in c74770f

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/FilterServlet.java:
```diff
@@ -60,23 +65,33 @@ public class FilterServlet extends SlingSafeMethodsServlet
         JsonObject allProperties = questionnaire == null
-            ? getAllFieldsFromAllQuestionnaires(request.getResourceResolver(), homepagePath)
-            : getAllFields(request.getResourceResolver(), questionnaire);
+            ? getAllFieldsFromAllQuestionnaires(request.getResourceResolver(), homepagePath, builder, includeQuestions)
```
These are methods whose name could start with `add`, as they are not returning anything and instead adding to the builder.

marta- replied:
Instead of passing `includeQuestions` here, we should just call this method inside an `if (includeQuestions) {}` in `doGet`.

marta- replied:
Done in c74770f

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/FilterServlet.java:
```diff
@@ -109,129 +127,119 @@ private JsonObject getAllFields(ResourceResolver resolver, String questionnaireP
-     */
-    private void copyQuestions(JsonObject questions, JsonObjectBuilder builder, Map<String, String> seenTypes,
-            Map<String, String> seenElements)
+    private JsonArrayBuilder addMetadataFilters(final boolean addQuestionnaireFilter)
```
Would be nice to have some javadoc, even if it's a private method.

marta- replied:
Done in c74770f

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/FilterServlet.java:
```diff
@@ -109,129 +127,119 @@ private JsonObject getAllFields(ResourceResolver resolver, String questionnaireP
-            if ("cards:Section".equals(datum.getString("jcr:primaryType"))) {
-                this.copyQuestions(datum, builder, seenTypes, seenElements);
-            }
+    private JsonObjectBuilder generateMetadataFilter(final String text)
```
Would be nice to have some javadoc, even if it's a private method.

marta- replied:
Done in c74770f

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/FilterServlet.java:
```diff
@@ -109,129 +127,119 @@ private JsonObject getAllFields(ResourceResolver resolver, String questionnaireP
-            // Copy over information from this object if this is a question
-            if ("cards:Question".equals(datum.getString("jcr:primaryType"))) {
-                copyFields(datum, key, builder, seenTypes, seenElements);
+    private void copyQuestionnaire(JsonObject resourceJson, JsonObjectBuilder questionnaire)
```
Would be nice to have some javadoc, even if it's a private method.

marta- replied:
I find the code easier to read if `questionnaire` is renamed to `questionnaireBuilder`.

marta- replied:
Addressed in c74770f

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/FilterServlet.java:
```diff
@@ -109,129 +127,119 @@ private JsonObject getAllFields(ResourceResolver resolver, String questionnaireP
+     * @param datum A JsonObject (from an cards:Questionnaire or cards:Section) whose fields may be cards:Questions
+     * @param questions questions
+     * @param sectionBreadcrumbs sectionBreadcrumbs
      * @return the content matching the query
```
It doesn't return anything anymore, javadoc needs update.

marta- replied:
Addressed in c74770f

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/FilterServlet.java:
```diff
@@ -109,129 +127,119 @@ private JsonObject getAllFields(ResourceResolver resolver, String questionnaireP
+        for (String key : question.keySet()) {
+            amended.add(key, question.get(key));
+        }
+        // Add the questionnaire title
```
Wrong comment, we're not adding the questionnaire title.

marta- replied:
Addressed in c74770f

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/FilterServlet.java:
```diff
@@ -109,129 +127,119 @@ private JsonObject getAllFields(ResourceResolver resolver, String questionnaireP
-     * @return the content matching the query
+     * @param question The question as a JsonObject
+     * @param sectionBreadcrumbs sectionBreadcrumbs
+     * @return a new JsonObject with all the fields of question, plus the questionnaireTitle field
```
Wrong comment, we're not adding the questionnaire title.

marta- replied:
Addressed in c74770f

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/FilterServlet.java:
```diff
@@ -109,129 +127,119 @@ private JsonObject getAllFields(ResourceResolver resolver, String questionnaireP
+                JsonObjectBuilder questionnaire = Json.createObjectBuilder();
+                JsonObject resourceJson = resource.adaptTo(JsonObject.class);
+                this.copyQuestionnaire(resourceJson, questionnaire);
+                builder.add(resourceJson.getString("@name"), questionnaire);
```
**Regression**: the filters are now grouped by questionnaire name instead of questionnaire title.
```suggestion
                builder.add(resourceJson.getString("title"), questionnaire);
```

marta- replied:
Addressed in c74770f

marta- committed:
CARDS-2171: Refactor FilterServlet to provide all the information needed by Filters.jsx
* Refactoring: reorganized code, reordered & renamed methods and variables, improved javadoc
* Bug fix: now displaying the correct breadcrumbs
* Regression fix: now grouping by questionnaire title instead of name
* Regression fix: now displaying the Questionnaire filter on dashboard

marta- committed:
CARDS-2171: Refactor FilterServlet to provide all the information needed by Filters.jsx
Updated FormView to accept a questionnairePath instead of a UUID and pass it on to the filters

marta- committed:
CARDS-2171: Refactor FilterServlet to provide all the information needed by Filters.jsx
Minor refactoring of the frontend filter data parsing

marta- committed:
CARDS-2171: Refactor FilterServlet to provide all the information needed by Filters.jsx
More comprehensive helper text for selected filters

sashaandjic reviewed (approved):
tested

sdumitriu committed:
CARDS-2171: Refactor FilterServlet to provide all the information needed by Filters.jsx
Bugfix: non-ASCII characters are broken
Bugfix: use the correct MIME type for the response

sdumitriu committed:
CARDS-2171: Refactor FilterServlet to provide all the information needed by Filters.jsx
Request filters directly from the questionnaire itself

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/FilterServlet.java:
```diff
@@ -60,178 +67,194 @@ public class FilterServlet extends SlingSafeMethodsServlet
-    private static final String JCR_UUID = "jcr:uuid";
-
     @Override
     public void doGet(final SlingHttpServletRequest request, final SlingHttpServletResponse response) throws IOException
```
Two related existing bugs here:
- the response is served as ASCII `text/plain` instead of `application/json`
- and thus is mangling all non-ASCII chars

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/FilterServlet.java:
```diff
@@ -60,178 +67,194 @@ public class FilterServlet extends SlingSafeMethodsServlet
 
```
This servlet can be applied to both the questionnaires homepage and a single questionnaire, but the implementation doesn't really support the latter case. It should be made to work, and instead of requesting `/Questionnaires.filters?questionnaire=/Questionnaires/OED`, the filters component should directly fetch `/Questionnaires/OED.filters`.
